Index: linux-2.6.21-gentoo-r2/drivers/ata/sata_via.c
===================================================================
--- linux-2.6.21-gentoo-r2.orig/drivers/ata/sata_via.c
+++ linux-2.6.21-gentoo-r2/drivers/ata/sata_via.c
@@ -120,10 +120,6 @@ static struct scsi_host_template svia_sh
 	.slave_configure	= ata_scsi_slave_config,
 	.slave_destroy		= ata_scsi_slave_destroy,
 	.bios_param		= ata_std_bios_param,
-#ifdef CONFIG_PM
-	.suspend		= ata_scsi_device_suspend,
-	.resume			= ata_scsi_device_resume,
-#endif
 };
 
 static const struct ata_port_operations vt6420_sata_ops = {
