Index: linux-git/drivers/net/Kconfig
===================================================================
--- linux-git.orig/drivers/net/Kconfig
+++ linux-git/drivers/net/Kconfig
@@ -2059,9 +2059,13 @@ config SKY2
 	  will be called sky2.  This is recommended.
 
 config SK98LIN
-	tristate "Marvell Yukon Chipset / SysKonnect SK-98xx Support"
+	tristate "Marvell Yukon Chipset / SysKonnect SK-98xx Support (DEPRECATED)"
 	depends on PCI
 	---help---
+	  This driver has been marked deprecated for Gentoo users, as it seems
+	  problematic and the new skge driver (CONFIG_SKGE) works much better.
+	  If you do not rely on features of this driver, please use SKGE instead.
+
 	  Say Y here if you have a Marvell Yukon or SysKonnect SK-98xx/SK-95xx
 	  compliant Gigabit Ethernet Adapter.
 
