diff -urN linux-2.6.37-tuxonice.orig/drivers/net/wireless/iwlwifi/iwl-5000.c linux-2.6.37-tuxonice/drivers/net/wireless/iwlwifi/iwl-5000.c
--- linux-2.6.37-tuxonice.orig/drivers/net/wireless/iwlwifi/iwl-5000.c	2011-01-05 03:50:19.000000000 +0300
+++ linux-2.6.37-tuxonice/drivers/net/wireless/iwlwifi/iwl-5000.c	2011-01-05 15:49:25.000000000 +0300
@@ -593,6 +593,7 @@
 	.mod_params = &iwlagn_mod_params,
 	.base_params = &iwl5000_base_params,
 	.ht_params = &iwl5000_ht_params,
+	.internal_wimax_coex = true,
 };
 
 struct iwl_cfg iwl5150_agn_cfg = {
@@ -610,6 +611,7 @@
 	.base_params = &iwl5000_base_params,
 	.ht_params = &iwl5000_ht_params,
 	.need_dc_calib = true,
+	.internal_wimax_coex = true,
 };
 
 struct iwl_cfg iwl5150_abg_cfg = {
@@ -626,6 +628,7 @@
 	.mod_params = &iwlagn_mod_params,
 	.base_params = &iwl5000_base_params,
 	.need_dc_calib = true,
+	.internal_wimax_coex = true,
 };
 
 MODULE_FIRMWARE(IWL5000_MODULE_FIRMWARE(IWL5000_UCODE_API_MAX));
diff -urN linux-2.6.37-tuxonice.orig/drivers/net/wireless/iwlwifi/iwl-6000.c linux-2.6.37-tuxonice/drivers/net/wireless/iwlwifi/iwl-6000.c
--- linux-2.6.37-tuxonice.orig/drivers/net/wireless/iwlwifi/iwl-6000.c	2011-01-05 03:50:19.000000000 +0300
+++ linux-2.6.37-tuxonice/drivers/net/wireless/iwlwifi/iwl-6000.c	2011-01-05 15:50:26.000000000 +0300
@@ -796,6 +796,7 @@
 	.base_params = &iwl6050_base_params,
 	.ht_params = &iwl6000_ht_params,
 	.need_dc_calib = true,
+	.internal_wimax_coex = true,
 };
 
 struct iwl_cfg iwl6050g2_bgn_cfg = {
@@ -830,6 +831,7 @@
 	.mod_params = &iwlagn_mod_params,
 	.base_params = &iwl6050_base_params,
 	.need_dc_calib = true,
+	.internal_wimax_coex = true,
 };
 
 struct iwl_cfg iwl6000_3agn_cfg = {
@@ -847,6 +849,7 @@
 	.base_params = &iwl6000_base_params,
 	.ht_params = &iwl6000_ht_params,
 	.need_dc_calib = true,
+	.internal_wimax_coex = true,
 };
 
 struct iwl_cfg iwl130_bgn_cfg = {
diff -urN linux-2.6.37-tuxonice.orig/drivers/net/wireless/iwlwifi/iwl-core.c linux-2.6.37-tuxonice/drivers/net/wireless/iwlwifi/iwl-core.c
--- linux-2.6.37-tuxonice.orig/drivers/net/wireless/iwlwifi/iwl-core.c	2011-01-05 03:50:19.000000000 +0300
+++ linux-2.6.37-tuxonice/drivers/net/wireless/iwlwifi/iwl-core.c	2011-01-05 15:49:25.000000000 +0300
@@ -1020,6 +1020,22 @@
 	/* Cancel currently queued command. */
 	clear_bit(STATUS_HCMD_ACTIVE, &priv->status);
 
+	/* W/A for WiFi/WiMAX coex and WiMAX own the RF */
+	if (priv->cfg->internal_wimax_coex &&
+			(!(iwl_read_prph(priv, APMG_CLK_CTRL_REG) &
+			   APMS_CLK_VAL_MRB_FUNC_MODE) ||
+			 (iwl_read_prph(priv, APMG_PS_CTRL_REG) &
+			  APMG_PS_CTRL_VAL_RESET_REQ))) {
+		wake_up_interruptible(&priv->wait_command_queue);
+		/*
+		 * Keep the restart process from trying to send host
+		 * commands by clearing the INIT status bit
+		 */
+		clear_bit(STATUS_READY, &priv->status);
+		IWL_ERR(priv, "RF is used by WiMAX\n");
+		return;
+	}
+
 	IWL_ERR(priv, "Loaded firmware version: %s\n",
 		priv->hw->wiphy->fw_version);
 
diff -urN linux-2.6.37-tuxonice.orig/drivers/net/wireless/iwlwifi/iwl-core.h linux-2.6.37-tuxonice/drivers/net/wireless/iwlwifi/iwl-core.h
--- linux-2.6.37-tuxonice.orig/drivers/net/wireless/iwlwifi/iwl-core.h	2011-01-05 03:50:19.000000000 +0300
+++ linux-2.6.37-tuxonice/drivers/net/wireless/iwlwifi/iwl-core.h	2011-01-05 15:49:25.000000000 +0300
@@ -344,6 +344,7 @@
  * @need_dc_calib: need to perform init dc calibration
  * @need_temp_offset_calib: need to perform temperature offset calibration
  * @scan_antennas: available antenna for scan operation
+ * @internal_wimax_coex: internal wifi/wimax combo device
  *
  * We enable the driver to be backward compatible wrt API version. The
  * driver specifies which APIs it supports (with @ucode_api_max being the
@@ -388,6 +389,7 @@
 	enum iwl_pa_type pa_type;	  /* if used set to IWL_PA_SYSTEM */
 	const bool need_dc_calib;	  /* if used set to true */
 	const bool need_temp_offset_calib; /* if used set to true */
+	const bool internal_wimax_coex;
 	u8 scan_rx_antennas[IEEE80211_NUM_BANDS];
 	u8 scan_tx_antennas[IEEE80211_NUM_BANDS];
 	const bool use_new_eeprom_reading; /* temporary, remove later */
diff -urN linux-2.6.37-tuxonice.orig/drivers/net/wireless/iwlwifi/iwl-prph.h linux-2.6.37-tuxonice/drivers/net/wireless/iwlwifi/iwl-prph.h
--- linux-2.6.37-tuxonice.orig/drivers/net/wireless/iwlwifi/iwl-prph.h	2011-01-05 03:50:19.000000000 +0300
+++ linux-2.6.37-tuxonice/drivers/net/wireless/iwlwifi/iwl-prph.h	2011-01-05 15:49:25.000000000 +0300
@@ -83,6 +83,7 @@
 #define APMG_DIGITAL_SVR_REG		(APMG_BASE + 0x0058)
 #define APMG_ANALOG_SVR_REG		(APMG_BASE + 0x006C)
 
+#define APMS_CLK_VAL_MRB_FUNC_MODE	(0x00000001)
 #define APMG_CLK_VAL_DMA_CLK_RQT	(0x00000200)
 #define APMG_CLK_VAL_BSM_CLK_RQT	(0x00000800)
 
