--- a/drivers/gpu/drm/i915/i915_gem.c	2010-12-17 09:59:07.000000000 -0500
+++ b/drivers/gpu/drm/i915/i915_gem.c	2010-12-17 09:59:51.000000000 -0500
@@ -2045,6 +2045,8 @@ i915_gpu_idle(struct drm_device *dev)
 	if (seqno1 == 0)
 		return -ENOMEM;
 	ret = i915_wait_request(dev, seqno1, &dev_priv->render_ring);
+	if (ret)
+		return ret;
 
 	if (HAS_BSD(dev)) {
 		seqno2 = i915_add_request(dev, NULL, I915_GEM_GPU_DOMAINS,
