diff -ruN linux-2.6.3.old/drivers/media/common/saa7146_core.c linux-2.6.3/drivers/media/common/saa7146_core.c
--- linux-2.6.3.old/drivers/media/common/saa7146_core.c	2004-02-17 22:58:47.000000000 -0500
+++ linux-2.6.3/drivers/media/common/saa7146_core.c	2004-02-21 14:34:35.432870456 -0500
@@ -69,14 +69,14 @@
 /* This DEBI code is based on the saa7146 Stradis driver by Nathan Laredo */
 int saa7146_wait_for_debi_done(struct saa7146_dev *dev)
 {
-	int start;
+	unsigned long start;
 
 	/* wait for registers to be programmed */
 	start = jiffies;
 	while (1) {
                 if (saa7146_read(dev, MC2) & 2)
                         break;
-		if (jiffies-start > HZ/20) {
+		if (time_after(jiffies, start + HZ/20)) {
 			DEB_S(("timed out while waiting for registers getting programmed\n"));
 			return -ETIMEDOUT;
 		}
@@ -88,7 +88,7 @@
 		if (!(saa7146_read(dev, PSR) & SPCI_DEBI_S))
 			break;
 		saa7146_read(dev, MC2);
-		if (jiffies-start > HZ/4) {
+		if (time_after(jiffies, start + HZ/4)) {
 			DEB_S(("timed out while waiting for transfer completion\n"));
 			return -ETIMEDOUT;
 		}
