--- a/arch/sparc/include/asm/tlb_64.h	2009-04-23 20:58:02.000000000 -0400
+++ b/arch/sparc/include/asm/tlb_64.h	2009-04-23 20:58:27.000000000 -0400
@@ -57,9 +57,9 @@ static inline struct mmu_gather *tlb_gat
 
 static inline void tlb_flush_mmu(struct mmu_gather *mp)
 {
+	if (!mp->fullmm)
+		flush_tlb_pending();
 	if (mp->need_flush) {
-		if (!mp->fullmm)
-			flush_tlb_pending();
 		free_pages_and_swap_cache(mp->pages, mp->pages_nr);
 		mp->pages_nr = 0;
 		mp->need_flush = 0;
