diff --git a/drivers/video/vesafb-tng.c b/drivers/video/vesafb-tng.c
index 0105a86..445701a 100644
--- a/drivers/video/vesafb-tng.c
+++ b/drivers/video/vesafb-tng.c
@@ -1561,8 +1561,8 @@ module_param_named(vgapal, pmi_setpal, i
 MODULE_PARM_DESC(vgapal,"bool: set palette using VGA registers");
 module_param_named(pmipal, pmi_setpal, bool, 0);
 MODULE_PARM_DESC(pmipal,"bool: set palette using PMI calls");
-module_param_named(nomtrr, mtrr, invbool, 0);
-MODULE_PARM_DESC(nomtrr,"bool: disable use of MTRR registers");
+module_param(mtrr, uint, 0);
+MODULE_PARM_DESC(mtrr,"Memory Type Range Registers setting. Use 0 to disable.");
 module_param(blank, bool, 1);
 MODULE_PARM_DESC(blank,"bool: enable hardware blanking");
 module_param(nocrtc, bool, 0);
