diff -ruN linux-2.6.7-hardened-r8.old/fs/Kconfig linux-2.6.7-hardened-r8/fs/Kconfig
--- linux-2.6.7-hardened-r8.old/fs/Kconfig	2004-10-13 10:18:40.000000000 -0400
+++ linux-2.6.7-hardened-r8/fs/Kconfig	2004-10-13 10:37:14.000000000 -0400
@@ -266,18 +266,6 @@
 
 	  If you don't know what Access Control Lists are, say N
 
-config REISERFS_FS_SECURITY
-	bool "ReiserFS Security Labels"
-	depends on REISERFS_FS_XATTR
-	help
-	  Security labels support alternative access control models
-	  implemented by security modules like SELinux.  This option
-	  enables an extended attribute handler for file security
-	  labels in the ReiserFS filesystem.
-
-	  If you are not using a security module that requires using
-	  extended attributes for file security labels, say N.
-
 config JFS_FS
 	tristate "JFS filesystem support"
 	select NLS
