--- linux/drivers/Makefile.orig	2005-09-12 14:18:27.000000000 +0100
+++ linux/drivers/Makefile	2005-09-12 14:18:42.000000000 +0100
@@ -24,6 +24,7 @@ obj-$(CONFIG_FB_INTEL)          += video
 obj-$(CONFIG_SERIO)		+= input/serio/
 obj-y				+= serial/
 obj-$(CONFIG_PARPORT)		+= parport/
+obj-$(CONFIG_GAMEPORT)		+= input/gameport/
 # load keyboard early so speakup can be quieted
 obj-$(CONFIG_INPUT)		+= input/
 obj-y				+= base/ block/ misc/ net/ media/
@@ -47,7 +48,6 @@ obj-$(CONFIG_PARIDE) 		+= block/paride/
 obj-$(CONFIG_TC)		+= tc/
 obj-$(CONFIG_USB)		+= usb/
 obj-$(CONFIG_USB_GADGET)	+= usb/gadget/
-obj-$(CONFIG_GAMEPORT)		+= input/gameport/
 obj-$(CONFIG_I2O)		+= message/
 obj-$(CONFIG_I2C)		+= i2c/
 obj-$(CONFIG_W1)		+= w1/
