--- a/drivers/net/wireless/iwlwifi/iwl-6000.c	2010-12-26 14:53:08.756994295 -0500
+++ b/drivers/net/wireless/iwlwifi/iwl-6000.c	2010-12-26 14:53:43.537393697 -0500
@@ -51,7 +51,7 @@
 
 /* Highest firmware API version supported */
 #define IWL6000_UCODE_API_MAX 4
-#define IWL6050_UCODE_API_MAX 4
+#define IWL6050_UCODE_API_MAX 5
 #define IWL6000G2_UCODE_API_MAX 4
 
 /* Lowest firmware API version supported */
