Index: linux/drivers/pci/proc.c
===================================================================
--- linux/drivers/pci/proc.c
+++ linux/drivers/pci/proc.c
@@ -572,7 +572,7 @@ static void legacy_proc_init(void)
 #ifdef CONFIG_GRKERNSEC_PROC_ADD
 #ifdef CONFIG_GRKERNSEC_PROC_USER
 	struct proc_dir_entry * entry = create_proc_entry("pci", S_IRUSR, NULL);
-#elif CONFIG_GRKERNSEC_PROC_USERGROUP
+#elif defined CONFIG_GRKERNSEC_PROC_USERGROUP
 	struct proc_dir_entry * entry = create_proc_entry("pci", S_IRUSR | S_IRGRP, NULL);
 #endif
 #else
@@ -609,7 +609,7 @@ static int __init pci_proc_init(void)
 #ifdef CONFIG_GRKERNSEC_PROC_ADD
 #ifdef CONFIG_GRKERNSEC_PROC_USER
 	proc_bus_pci_dir = proc_mkdir_mode("pci", S_IRUSR | S_IXUSR, proc_bus);
-#elif CONFIG_GRKERNSEC_PROC_USERGROUP
+#elif defined CONFIG_GRKERNSEC_PROC_USERGROUP
 	proc_bus_pci_dir = proc_mkdir_mode("pci", S_IRUSR | S_IXUSR | S_IRGRP | S_IXGRP, proc_bus);
 #endif
 #else
Index: linux/fs/proc/base.c
===================================================================
--- linux/fs/proc/base.c
+++ linux/fs/proc/base.c
@@ -2045,7 +2045,7 @@ struct dentry *proc_pid_lookup(struct in
 
 #ifdef CONFIG_GRKERNSEC_PROC_USER
 	inode->i_mode = S_IFDIR|S_IRUSR|S_IXUSR;
-#elif CONFIG_GRKERNSEC_PROC_USERGROUP
+#elif defined CONFIG_GRKERNSEC_PROC_USERGROUP
 	inode->i_mode = S_IFDIR|S_IRUSR|S_IXUSR|S_IRGRP|S_IXGRP;
 	inode->i_gid = CONFIG_GRKERNSEC_PROC_GID;
 #else
Index: linux/fs/proc/proc_misc.c
===================================================================
--- linux/fs/proc/proc_misc.c
+++ linux/fs/proc/proc_misc.c
@@ -737,7 +737,7 @@ void __init proc_misc_init(void)
 
 #ifdef CONFIG_GRKERNSEC_PROC_USER
 	gr_mode = S_IRUSR;
-#elif CONFIG_GRKERNSEC_PROC_USERGROUP
+#elif defined CONFIG_GRKERNSEC_PROC_USERGROUP
 	gr_mode = S_IRUSR | S_IRGRP;
 #endif
 #ifdef CONFIG_GRKERNSEC_PROC_ADD
Index: linux/fs/proc/root.c
===================================================================
--- linux/fs/proc/root.c
+++ linux/fs/proc/root.c
@@ -55,7 +55,7 @@ void __init proc_root_init(void)
 	proc_misc_init();
 #ifdef CONFIG_GRKERNSEC_PROC_USER
 	proc_net = proc_mkdir_mode("net", S_IRUSR | S_IXUSR, NULL);
-#elif CONFIG_GRKERNSEC_PROC_USERGROUP
+#elif defined CONFIG_GRKERNSEC_PROC_USERGROUP
 	proc_net = proc_mkdir_mode("net", S_IRUSR | S_IXUSR | S_IRGRP | S_IXGRP, NULL);
 #else
 	proc_net = proc_mkdir("net", NULL);
@@ -86,7 +86,7 @@ void __init proc_root_init(void)
 #ifdef CONFIG_GRKERNSEC_PROC_ADD
 #ifdef CONFIG_GRKERNSEC_PROC_USER
 	proc_bus = proc_mkdir_mode("bus", S_IRUSR | S_IXUSR, NULL);
-#elif CONFIG_GRKERNSEC_PROC_USERGROUP
+#elif defined CONFIG_GRKERNSEC_PROC_USERGROUP
 	proc_bus = proc_mkdir_mode("bus", S_IRUSR | S_IXUSR | S_IRGRP | S_IXGRP, NULL);
 #endif
 #else
Index: linux/kernel/configs.c
===================================================================
--- linux/kernel/configs.c
+++ linux/kernel/configs.c
@@ -92,7 +92,7 @@ static int __init ikconfig_init(void)
 #ifdef CONFIG_GRKERNSEC_PROC_ADD
 #ifdef CONFIG_GRKERNSEC_PROC_USER
 	entry = create_proc_entry("config.gz", S_IFREG | S_IRUSR, &proc_root);
-#elif CONFIG_GRKERNSEC_PROC_USERGROUP
+#elif defined CONFIG_GRKERNSEC_PROC_USERGROUP
 	entry = create_proc_entry("config.gz", S_IFREG | S_IRUSR | S_IRGRP, &proc_root);
 #endif
 #else
Index: linux/kernel/kallsyms.c
===================================================================
--- linux/kernel/kallsyms.c
+++ linux/kernel/kallsyms.c
@@ -413,7 +413,7 @@ static int __init kallsyms_init(void)
 #ifdef CONFIG_GRKERNSEC_PROC_ADD
 #ifdef CONFIG_GRKERNSEC_PROC_USER
 	entry = create_proc_entry("kallsyms", S_IFREG | S_IRUSR, NULL);
-#elif CONFIG_GRKERNSEC_PROC_USERGROUP
+#elif defined CONFIG_GRKERNSEC_PROC_USERGROUP
 	entry = create_proc_entry("kallsyms", S_IFREG | S_IRUSR | S_IRGRP, NULL);
 #endif
 #else
Index: linux/kernel/resource.c
===================================================================
--- linux/kernel/resource.c
+++ linux/kernel/resource.c
@@ -139,7 +139,7 @@ static int __init ioresources_init(void)
 #ifdef CONFIG_GRKERNSEC_PROC_ADD
 #ifdef CONFIG_GRKERNSEC_PROC_USER
 	entry = create_proc_entry("ioports", S_IRUSR, NULL);
-#elif CONFIG_GRKERNSEC_PROC_USERGROUP
+#elif defined CONFIG_GRKERNSEC_PROC_USERGROUP
 	entry = create_proc_entry("ioports", S_IRUSR | S_IRGRP, NULL);
 #endif
 #else
@@ -151,7 +151,7 @@ static int __init ioresources_init(void)
 #ifdef CONFIG_GRKERNSEC_PROC_ADD
 #ifdef CONFIG_GRKERNSEC_PROC_USER
 	entry = create_proc_entry("iomem", S_IRUSR, NULL);
-#elif CONFIG_GRKERNSEC_PROC_USERGROUP
+#elif defined CONFIG_GRKERNSEC_PROC_USERGROUP
 	entry = create_proc_entry("iomem", S_IRUSR | S_IRGRP, NULL);
 #endif
 #else
