--- a/drivers/gpu/drm/i915/intel_tv.c	2013-10-01 19:07:03.715400887 -0400
+++ b/drivers/gpu/drm/i915/intel_tv.c	2013-10-01 19:08:34.635400020 -0400
@@ -918,6 +918,14 @@ intel_tv_compute_config(struct intel_enc
 	DRM_DEBUG_KMS("forcing bpc to 8 for TV\n");
 	pipe_config->pipe_bpp = 8*3;
 
+	/* TV has it's own notion of sync and other mode flags, so clear them. */
+	pipe_config->adjusted_mode.flags = 0;
+
+	/*
+	 * FIXME: We don't check whether the input mode is actually what we want
+	 * or whether userspace is doing something stupid.
+	 */
+
 	return true;
 }
 
