--- linux-2.6.17-gentoo-r2/drivers/Makefile	2006-07-10 10:51:59.000000000 +0200
+++ linux-2.6.17-gentoo-r2-local-2/drivers/Makefile	2006-07-10 12:24:07.000000000 +0200
@@ -26,6 +26,7 @@
 
 obj-y				+= serial/
 obj-$(CONFIG_PARPORT)		+= parport/
+obj-$(CONFIG_SERIO)		+= input/serio/
 obj-$(CONFIG_GAMEPORT)		+= input/gameport/
 # load keyboard early so speakup can be quieted
 obj-$(CONFIG_INPUT)		+= input/
@@ -52,7 +53,6 @@
 obj-$(CONFIG_USB)		+= usb/
 obj-$(CONFIG_PCI)		+= usb/
 obj-$(CONFIG_USB_GADGET)	+= usb/gadget/
-obj-$(CONFIG_SERIO)		+= input/serio/
 obj-$(CONFIG_I2O)		+= message/
 obj-$(CONFIG_RTC_LIB)		+= rtc/
 obj-$(CONFIG_I2C)		+= i2c/
