Backported to 2.6.14.7 by Kerin Millar <kerframil@gmail.com>:

From stable-bounces@linux.kernel.org  Wed Feb  1 21:18:58 2006
Date: Thu, 02 Feb 2006 00:09:00 -0500
From: Dmitry Torokhov <dtor_core@ameritech.net>
To: stable@kernel.org
Cc: 
Subject: Input: grip - fix crash when accessing device

Signed-off-by: Dmitry Torokhov <dtor@mail.ru>
Signed-off-by: Chris Wright <chrisw@sous-sol.org>
---

--- linux-2.6.14.7.orig/drivers/input/joystick/grip.c	2005-10-27 19:02:08.000000000 -0500
+++ linux-2.6.14.7/drivers/input/joystick/grip.c	2006-02-06 21:12:44.000000000 -0500
@@ -191,6 +191,9 @@ static void grip_poll(struct gameport *g
 	for (i = 0; i < 2; i++) {
 
 		dev = grip->dev + i;
+		if (!dev)
+			continue;
+
 		grip->reads++;
 
 		switch (grip->mode[i]) {
