--- linux-2.6.13-gentoo-r4/include/sound/pcm.h.orig	2005-10-25 20:56:35.000000000 +0100
+++ linux-2.6.13-gentoo-r4/include/sound/pcm.h	2005-10-25 20:59:11.000000000 +0100
@@ -904,6 +904,7 @@ int snd_pcm_format_unsigned(snd_pcm_form
 int snd_pcm_format_linear(snd_pcm_format_t format);
 int snd_pcm_format_little_endian(snd_pcm_format_t format);
 int snd_pcm_format_big_endian(snd_pcm_format_t format);
+#if 0 /* just for docbook */
 /**
  * snd_pcm_format_cpu_endian - Check the PCM format is CPU-endian
  * @format: the format to check
@@ -911,7 +912,8 @@ int snd_pcm_format_big_endian(snd_pcm_fo
  * Returns 1 if the given PCM format is CPU-endian, 0 if
  * opposite, or a negative error code if endian not specified.
  */
-/* int snd_pcm_format_cpu_endian(snd_pcm_format_t format); */
+int snd_pcm_format_cpu_endian(snd_pcm_format_t format);
+#endif /* docbook */
 #ifdef SNDRV_LITTLE_ENDIAN
 #define snd_pcm_format_cpu_endian	snd_pcm_format_little_endian
 #else
