From unknown

Workaround for the Ultra1E onboard HME lockup.

-------------------------------------------------------------------------------

Index: linux-2.6.18/drivers/net/sunhme.c
===================================================================
--- linux-2.6.18.orig/drivers/net/sunhme.c
+++ linux-2.6.18/drivers/net/sunhme.c
@@ -1971,6 +1971,7 @@ static void happy_meal_tx(struct happy_m
 	}
 	hp->tx_old = elem;
 	TXD((">"));
+	udelay(1);
 
 	if (netif_queue_stopped(dev) &&
 	    TX_BUFFS_AVAIL(hp) > (MAX_SKB_FRAGS + 1))
