--- a/drivers/gpu/drm/drm_crtc.c	2010-09-29 10:10:59.000000000 -0400
+++ b/drivers/gpu/drm/drm_crtc.c	2010-09-29 10:12:56.000000000 -0400
@@ -2609,6 +2609,15 @@ int drm_mode_page_flip_ioctl(struct drm_
 	if (!obj)
 		goto out;
 	crtc = obj_to_crtc(obj);
+ 
+	if (crtc->fb == NULL) {
+		/* The framebuffer is currently unbound, presumably
+		 * due to a hotplug event, that userspace has not
+		 * yet discovered.
+		 */
+		ret = -EBUSY;
+		goto out;
+	}
 
 	if (crtc->funcs->page_flip == NULL)
 		goto out;
