--- a/drivers/net/wireless/iwlwifi/iwl-agn-sta.c	2012-06-13 12:14:13.444172356 -0400
+++ b/drivers/net/wireless/iwlwifi/iwl-agn-sta.c	2012-06-13 12:15:19.223797305 -0400
@@ -1222,7 +1222,7 @@ int iwl_remove_dynamic_key(struct iwl_pr
 		key_flags |= STA_KEY_MULTICAST_MSK;
 
 	sta_cmd.key.key_flags = key_flags;
-	sta_cmd.key.key_offset = WEP_INVALID_OFFSET;
+    sta_cmd.key.key_offset = keyconf->hw_key_idx;
 	sta_cmd.sta.modify_mask = STA_MODIFY_KEY_MASK;
 	sta_cmd.mode = STA_CONTROL_MODIFY_MSK;
 
