partial fixup of linux-2.6.27-s390-01-007-zfcp-perf-data.diff

--- linux-2.6.27/drivers/s390/scsi/zfcp_fsf.c
+++ linux-2.6.27/drivers/s390/scsi/zfcp_fsf.c
@@ -782,7 +782,7 @@
 	list_add_tail(&req->list, &adapter->req_list[idx]);
 	spin_unlock_irqrestore(&adapter->req_list_lock, flags);
 
-	req->qdio_outb_usage = atomic_read(&req_q->count);
+	req->qdio_outb_usage = atomic_read(&adapter->req_q.count);
 	req->issued = get_clock();
 	if (zfcp_qdio_send(req)) {
 		del_timer(&req->timer);
