--- linux/fs/exec.c	2005-06-14 23:05:42.000000000 +0100
+++ linux/fs/exec.c	2005-06-14 23:07:46.000000000 +0100
@@ -1618,9 +1618,11 @@
 		}
 		up_read(&mm->mmap_sem);
 	}
+#ifdef CONFIG_GRKERNSEC_PROC_IPADDR
 	if (tsk->signal->curr_ip)
 		printk(KERN_ERR "PAX: From %u.%u.%u.%u: execution attempt in: %s, %08lx-%08lx %08lx\n", NIPQUAD(tsk->signal->curr_ip), path_fault, start, end, offset);
 	else
+#endif
 		printk(KERN_ERR "PAX: execution attempt in: %s, %08lx-%08lx %08lx\n", path_fault, start, end, offset);
 	printk(KERN_ERR "PAX: terminating task: %s(%s):%d, uid/euid: %u/%u, "
 			"PC: %p, SP: %p\n", path_exec, tsk->comm, tsk->pid,
diff -Nur linux-2.6.11-grsec/security/Kconfig linux-2.6.11-grsec.lorenzo/security/Kconfig
--- linux/security/Kconfig	2005-03-09 22:30:09.000000000 +0100
+++ linux/security/Kconfig	2005-03-09 22:32:10.000000000 +0100
@@ -10,7 +10,7 @@
 
 config PAX
 	bool "Enable various PaX features"
-	depends on GRKERNSEC && (ALPHA || ARM || IA64 || MIPS32 || MIPS64 || PARISC || PPC32 || PPC64 || SPARC32 || SPARC64 || X86 || X86_64)
+	depends on (ALPHA || ARM || IA64 || MIPS32 || MIPS64 || PARISC || PPC32 || PPC64 || SPARC32 || SPARC64 || X86 || X86_64)
 	help
 	  This allows you to enable various PaX features.  PaX adds
 	  intrusion prevention mechanisms to the kernel that reduce
--- linux/arch/i386/mm/fault.c.orig	2005-09-16 18:56:11.000000000 +0100
+++ linux/arch/i386/mm/fault.c	2005-09-16 18:57:26.000000000 +0100
@@ -610,10 +610,12 @@
 #else
 	else if (init_mm.start_code <= address && address < init_mm.end_code)
 #endif
+#ifdef CONFIG_GRKERNSEC_PROC_IPADDR
 		if (tsk->signal->curr_ip)
 			printk(KERN_ERR "PAX: From %u.%u.%u.%u: %s:%d, uid/euid: %u/%u, attempted to modify kernel code",
 					 NIPQUAD(tsk->signal->curr_ip), tsk->comm, tsk->pid, tsk->uid, tsk->euid);
 		else
+#endif
 			printk(KERN_ERR "PAX: %s:%d, uid/euid: %u/%u, attempted to modify kernel code",
 					 tsk->comm, tsk->pid, tsk->uid, tsk->euid);
 #endif
